src/Castle.Windsor/MicroKernel/DefaultKernel.cs

```diff
@@ -278,9 +279,22 @@ IHandler IKernelInternal.CreateHandler(ComponentModel model)
 
 		public virtual IKernel AddFacility(IFacility facility)
 		{
-#pragma warning disable 612,618
-			return AddFacility(facility != null ? facility.GetType().FullName : null, facility);
-#pragma warning restore 612,618
+			if (facility == null)
+			{
+				throw new ArgumentNullException("facility");
+			}
+			var facilityType = facility.GetType();
+			if (facilities.Any(f => f.GetType() == facilityType))
+			{
+				throw new ArgumentException(
+					string.Format(
+						"Facility of type '{0}' has already been registered with the container. Only one facility of a given type can exist in the container.",
+						facilityType.FullName));
+			}
+			facilities.Add(facility);
+			facility.Init(this, ConfigurationStore.GetFacilityConfiguration(facility != null ? facility.GetType().FullName : null));
```

Comment:
`facility` can't be null here.

At first this appears to be a breaking change to loading of facility configuration from an XML file because you've removed the ability to define the "id" attribute which both the code and XML share, however that breaking change was actually made in 3.0.0 according to our changelog:

> change - type attribute is now required and id is ignored in facility XML configuration
> impact - low
> fixability - easy
> description - Since type is uniquely identifying facilities there was no point in keeping the id
> 	around anymore.
> fix - This change can affect you in two ways. If you were using facilities node in the XML and
> 	not specifying the type it is now mandatory. Notice Windsor's ability to apply short type
> 	names works here as well, so often just type name is enough - no need to specify assembly
> 	qualified name. Also the assembly will now be instantiated by the container, so if you were
> 	adding it in code later on, this is no longer required (in fact it will throw an exception
> 	saying the assembly was already added).
> 	The other thing that may affect you is if you were looking up facility config namnually via
> 	IConfigurationStore.GetFacilityConfiguration method. It now expects full name of the type
> 	as the key, so you should be calling it like this:
> 	store.GetFacilityConfiguration(typeof(YourFacility).FullName);

What we should actually do here is remove any mention of the `id` attribute in our docs, XML docs and unit tests since that attribute is ignored, `DefaultConfigurationStore.facilities` holds a map of type full name to facility instance.

`XmlInterpreter ` and `DefaultComponentInstaller` definitely ignore `id`, I'm thinking that it should actually throw on the existence of the attribute now that so much time has passed and because it just confuses people that the `id` attribute is still actually used.

What do you think?

Reply:
Yup, dealt with. I updated it everywhere.

Reply:
Thanks.

Did you have an opinion on my last part of that comment, throwing if someone does actually specify an `id` attribute to force people to remove them when they upgrade to clean up their XML configuration? Happy if you think it is more work than it is worth, especially that you've just removed the obsolete `AddFacility` methods so users should realise the id/key is ignored if they were using those ones, otherwise there is no harm.